### PR TITLE
added stats page for referrals (shared, and received)

### PR DESCRIPTION
**Pull request by yenicelik**
- added stats for showing how much a user has gained from referrals
- added stats for used up referrals

### CODE REVIEW

**Comment by yenicelik** [comment]
will still have to write a short test to test it out, but just checking with the PR if there's anything i'm missing @WyseNynja

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:106]: I'm wondering if we actually want to expose the address. Just because I refer you doesn't mean I want you to be able to track me on chain.

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:160]: I'm not sure we want an error here. An empty response seems better.

File: web3_proxy/src/frontend/users/referral.rs
Comment (yenicelik) [reply web3_proxy/src/frontend/users/referral.rs:106]: I had the same thought, I will remove this for now then

**Comment by yenicelik** [comment]
I still have to test it, please only approve (dont merge) if ready

File: web3_proxy/src/frontend/mod.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/mod.rs:195]: its not a huge deal, but "used" and "shared" seem like odd terms. I don't have any better suggestions though.

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:96]: `conn` -> `as_ref`

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:103]: lets keep them decimal for now

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:110]: is unwrap needed/safe here?

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:153]: `conn` -> `as_ref`

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:159]: no need to allocate a vec here. `[]` should be fine

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:160]: do we want `""`, or would it be better to do `None` (which json will turn into `null`)?

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:172]: same as above, lets keep Decimal for now

File: web3_proxy/src/frontend/users/referral.rs
Comment (WyseNynja) [inline web3_proxy/src/frontend/users/referral.rs:191]: no need for the `format!`. `Address` already serializes how you want it

Comment (WyseNynja) [reply web3_proxy/src/frontend/users/referral.rs:191]: I also mentioned this on another PR. Is it necessary to expose the on-chain address here? I'd prefer to keep that private as much as possible.

**Review by WyseNynja — CHANGES_REQUESTED** [review]

File: web3_proxy/src/frontend/users/referral.rs
Comment (yenicelik) [reply web3_proxy/src/frontend/users/referral.rs:103]: ok cool

File: web3_proxy/src/frontend/users/referral.rs
Comment (yenicelik) [reply web3_proxy/src/frontend/users/referral.rs]: done

File: web3_proxy/src/frontend/mod.rs
Comment (yenicelik) [reply web3_proxy/src/frontend/mod.rs:200]: just changed it to `used-codes` and `shared-codes`? still not amazing but better

File: web3_proxy/src/frontend/users/referral.rs
Comment (yenicelik) [reply web3_proxy/src/frontend/users/referral.rs:110]: each referral_record comes with a referrer, technically it should be (each referral code has a referrer). i can modify it to use context though

Comment (yenicelik) [reply web3_proxy/src/frontend/users/referral.rs]: changed it to context

File: web3_proxy/src/frontend/users/referral.rs
Comment (yenicelik) [reply web3_proxy/src/frontend/users/referral.rs:159]: changed

File: web3_proxy/src/frontend/users/referral.rs
Comment (yenicelik) [reply web3_proxy/src/frontend/users/referral.rs:172]: removed comment

File: web3_proxy/src/frontend/users/referral.rs
Comment (yenicelik) [reply web3_proxy/src/frontend/users/referral.rs]: I removed it. Was assuming that the referrer would know who to share the address with. Actually, he does know when making the referral, so I'll keep this in. Let me know if I should remove it though.

File: web3_proxy/src/frontend/users/referral.rs
Comment (yenicelik) [reply web3_proxy/src/frontend/users/referral.rs]: changed

**Comment by yenicelik** [comment]
Only this is basically left for discussion, https://github.com/llamanodes/web3-proxy/pull/95#discussion_r1210941230 let me know, should be ready for merge after testing then

**Review by WyseNynja — APPROVED** [review]

**Comment by yenicelik** [comment]
I will do some testing once we push things to devel if this is ok? I don't want to write redundant tests right now haha (happy to do so though if you prefer so)

**Pull request merged**
